src/libs/antares/study/parallel-years.cpp

```diff
@@ -0,0 +1,431 @@
+        break;
+    case 3:
+        numberOfMCYearThreads[ncLow] = 2;
+        numberOfMCYearThreads[ncAvg] = 2;
+        numberOfMCYearThreads[ncHigh] = 2;
+        numberOfMCYearThreads[ncMax] = 3;
+        break;
+    case 4:
+        numberOfMCYearThreads[ncLow] = 2;
+        numberOfMCYearThreads[ncAvg] = 2;
+        numberOfMCYearThreads[ncHigh] = 3;
+        numberOfMCYearThreads[ncMax] = 4;
+        break;
+    case 5:
+        numberOfMCYearThreads[ncLow] = 2;
+        numberOfMCYearThreads[ncAvg] = 3;
+        numberOfMCYearThreads[ncHigh] = 4;
+        numberOfMCYearThreads[ncMax] = 5;
+        break;
+    case 6:
+        numberOfMCYearThreads[ncLow] = 2;
+        numberOfMCYearThreads[ncAvg] = 3;
+        numberOfMCYearThreads[ncHigh] = 4;
+        numberOfMCYearThreads[ncMax] = 6;
+        break;
+    case 7:
+        numberOfMCYearThreads[ncLow] = 2;
+        numberOfMCYearThreads[ncAvg] = 3;
+        numberOfMCYearThreads[ncHigh] = 5;
+        numberOfMCYearThreads[ncMax] = 7;
+        break;
+    case 8:
+        numberOfMCYearThreads[ncLow] = 2;
+        numberOfMCYearThreads[ncAvg] = 4;
+        numberOfMCYearThreads[ncHigh] = 6;
+        numberOfMCYearThreads[ncMax] = 8;
+        break;
+    case 9:
+        numberOfMCYearThreads[ncLow] = 3;
+        numberOfMCYearThreads[ncAvg] = 5;
+        numberOfMCYearThreads[ncHigh] = 7;
+        numberOfMCYearThreads[ncMax] = 8;
+        break;
+    case 10:
+        numberOfMCYearThreads[ncLow] = 3;
+        numberOfMCYearThreads[ncAvg] = 5;
+        numberOfMCYearThreads[ncHigh] = 8;
+        numberOfMCYearThreads[ncMax] = 9;
+        break;
+    case 11:
+        numberOfMCYearThreads[ncLow] = 3;
+        numberOfMCYearThreads[ncAvg] = 6;
+        numberOfMCYearThreads[ncHigh] = 8;
+        numberOfMCYearThreads[ncMax] = 10;
+        break;
+    case 12:
+        numberOfMCYearThreads[ncLow] = 3;
+        numberOfMCYearThreads[ncAvg] = 6;
+        numberOfMCYearThreads[ncHigh] = 9;
+        numberOfMCYearThreads[ncMax] = 11;
+        break;
+    default:
+        numberOfMCYearThreads[ncLow] = (uint)std::ceil(nbLogicalCores / 4.);
+        numberOfMCYearThreads[ncAvg] = (uint)std::ceil(nbLogicalCores / 2.);
+        numberOfMCYearThreads[ncHigh] = (uint)std::ceil(3 * nbLogicalCores / 4.);
+        numberOfMCYearThreads[ncMax] = nbLogicalCores - 1;
+        break;
+    }
+
+    /*
+            Getting the number of parallel years based on the number
+            of cores level.
+            This number is limited by the smallest refresh span (if at least
+            one type of time series is generated)
+    */
+
+    try
+    {
+        forcedNbOfParallelYears_ = numberOfMCYearThreads.at(p.nbCores.ncMode);
+    }
+    catch(const std::out_of_range& e)
+    {
+        logs.fatal() << "Simulation cores level not correct : " << (int)p.nbCores.ncMode;
+    }
+    
+}
+
+
+void SetsOfParallelYearCalculator::limitNbOfParallelYearsbyMinRefreshSpan()
```

Comment:
About functions **limitNbOfParallelYearsbyMinRefreshSpan** and **isRefreshNeededForCurrentYear** : 
- there are some duplication code in both functions
- they imply a dependency of class **SetsOfParallelYearCalculator** to **study->parameter**, making this class more difficult to test in isolation that it should.

Just a first think (to be digged)
So we could have a parent class **refreshSpanHandler** : 
```C++
class refreshSpanHandler
{
     getSpan() = 0;
     doWeRefresh(int year) = 0;
}
```
Then derive it for all kinds of production or load.
Maybe build them (create instance of them) outside the class **SetsOfParallelYearCalculator**.
Add them to the parallel years calculator, that is put them in a vector of parent class **refreshSpanHandler**, let's say **refreshSpanHandlerList**)
Then we could traverse this vector with an std::algorithm.
example : 
```C++
 refreshing = std::any_of(refreshSpanHandlerList.begin(), refreshSpanHandlerList.end(), [](const refreshSpanHandler h) { return h->doWeRefresh(y)});
```